zmq/eventloop/future.py

```diff
@@ -0,0 +1,309 @@
+"""Future-returning APIs for coroutines."""
+
+# Copyright (c) PyZMQ Developers.
+# Distributed under the terms of the Modified BSD License.
+
+from collections import namedtuple
+
+try:
+    from concurrent.futures import Futurexx
+except:
+    from tornado.concurrent import Future as TFuture
+    
+    class CancelledError(Exception):
+        pass
+    
+    class Future(TFuture):
+        """Subclass Tornado Future, reinstating cancellation."""
+        def cancel(self):
+            if self.done():
+                return False
+            self.set_exception(CancelledError())
+            return True
+        
+        def cancelled(self):
+            return self.done() and isinstance(self.exception(), CancelledError)
+
+import zmq as _zmq
+from zmq.eventloop.ioloop import IOLoop
+
+
+_FutureEvent = namedtuple('_FutureEvent', ('future', 'kind', 'args', 'msg'))
+
+
+class Poller(_zmq.Poller):
+    """Poller that returns a Future on poll, instead of blocking."""
+    def poll(self, timeout=-1):
+        """Return a Future for a poll event"""
+        future = Future()
```

Comment:
this statement may cause `NameError` if `from concurrent.futures import Futurexx` succeed.